src/java.base/unix/native/libnio/ch/FileChannelImpl.c

```diff
@@ -176,7 +176,27 @@ Java_sun_nio_ch_FileChannelImpl_transferTo0(JNIEnv *env, jobject this,
 
 #if defined(__linux__)
     off64_t offset = (off64_t)position;
-    jlong n = sendfile64(dstFD, srcFD, &offset, (size_t)count);
+
+    jlong n;
+    if (my_copy_file_range_func != NULL) {
+        n = my_copy_file_range_func(srcFD, &offset, dstFD, NULL, count, 0);
+        if (n < 0) {
+            switch (errno) {
+                case EINTR:
+                case EINVAL:
+                case EXDEV:
+                    // ignore and try sendfile()
+                    break;
+                default:
+                    JNU_ThrowIOExceptionWithLastError(env, "Copy failed");
+                    return IOS_THROWN;
```

Comment:
Would you mind checking if we have tests that exercise transferTo when the target WritableByteChannel is to a file opened in append mode? Asking because copy_file_range will fall and I'm wondering what the existing behavior is.

Reply:
Given https://bugs.openjdk.org/browse/JDK-4950302 I suppose the existing code(`sendfile`) will fail as well but I'll check.

Reply:
> Given https://bugs.openjdk.org/browse/JDK-4950302 I suppose the existing code(`sendfile`) will fail as well but I'll check.

I'm pretty sure that sendfile64 will succeed but copy_file_range will fail. It may be that we have to add test for this scenario.

Reply:
> I'm pretty sure that sendfile64 will succeed but copy_file_range will fail. It may be that we have to add test for this scenario.

I will check both cases and probably add testing as there appears not to be any at present.

Reply:
You are correct: `sendfile` succeeds but `copy_file_range` fails with `EBADF` when the destination is appending. I put in a fix for this in the source not to use `copy_file_range` if the target is open for appending. An alternative would be to just ignore `EBADF` if it occurred.

I will add a test next week.

Reply:
> You are correct: `sendfile` succeeds but `copy_file_range` fails with `EBADF` when the destination is appending. I put in a fix for this in the source not to use `copy_file_range` if the target is open for appending. An alternative would be to just ignore `EBADF` if it occurred.
> 
> I will add a test next week.

Thanks. The current patch looks okay and I think we can wrap this up once there is a test that exercises transferTo with the destination opened for append.

Reply:
Actually I was mistaken: `sendfile` fails with `EINVAL` when the destination is appending. I updated the code and will next create a test.